### ISSUE TITLE
index out of range panic on deposits via CLI

### ISSUE THREAD

**Issue by jtieri** [duality-labs/duality#291]
## Description
I was cleaning up the IBC swap middleware implementation and moving the test cases into the repo when I noticed that attempting to submit a deposit via the CLI yields a panic due to an index out of range error.

It looks like `MsgDeposit` recently had a new field added to it, `Options     []*DepositOptions`
In `CmdDeposit` there is not an expected argument for this new field and nothing in the constructor explicitly initializes the slice https://github.com/duality-labs/duality/blob/f950d8c71b0ad76214c09d71e033bb60f36b3f17/x/dex/client/cli/tx_deposit.go#L81-L90

So the msg is composed and the tx gets broadcasted, then in the msg server where the deposit is handled there is an assumption that this slice is equal in length to the `amounts0` slice https://github.com/duality-labs/duality/blob/f950d8c71b0ad76214c09d71e033bb60f36b3f17/x/dex/keeper/core.go#L46

The issue goes unnoticed before the tx is broadcasted because `ValidateBasic` does not assert that the `Options` slice is equal in length to the other slices in `MsgDeposit`

https://github.com/duality-labs/duality/blob/f950d8c71b0ad76214c09d71e033bb60f36b3f17/x/dex/types/message_deposit.go#L58-L62

Here's the stack trace:
```
exit code 1:  Error: rpc error: code = Unknown desc = rpc error: code = Unknown desc = recovered: runtime error: index out of range [0] with length 0
        	            	stack:
        	            	goroutine 2239 [running]:
        	            	runtime/debug.Stack()
        	            		runtime/debug/stack.go:24 +0x64
        	            	github.com/cosmos/cosmos-sdk/baseapp.newDefaultRecoveryMiddleware.func1({0x189cd60, 0x4004d29380})
        	            		github.com/cosmos/cosmos-sdk@v0.46.6/baseapp/recovery.go:71 +0x24
        	            	github.com/cosmos/cosmos-sdk/baseapp.newRecoveryMiddleware.func1({0x189cd60?, 0x4004d29380?})
        	            		github.com/cosmos/cosmos-sdk@v0.46.6/baseapp/recovery.go:39 +0x34
        	            	github.com/cosmos/cosmos-sdk/baseapp.processRecovery({0x189cd60, 0x4004d29380}, 0x400022a0c0?)
        	            		github.com/cosmos/cosmos-sdk@v0.46.6/baseapp/recovery.go:28 +0x38
        	            	github.com/cosmos/cosmos-sdk/baseapp.processRecovery({0x189cd60, 0x4004d29380}, 0x244da20?)
        	            		github.com/cosmos/cosmos-sdk@v0.46.6/baseapp/recovery.go:33 +0x60
        	            	github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).runTx.func1()
        	            		github.com/cosmos/cosmos-sdk@v0.46.6/baseapp/baseapp.go:632 +0xb0
        	            	panic({0x189cd60, 0x4004d29380})
        	            		runtime/panic.go:884 +0x20c
        	            	github.com/NicholasDotSol/duality/x/dex/keeper.Keeper.DepositCore({{0x244be50, 0x4000f5c8b0}, {0x242a618, 0x4000fde4d0}, {0x242a618, 0x0}, {{0x244be50, 0x4000f5c8b0}, 0x40001a02e0, {0x242a618, ...}, ...}, ...}, ...)
        	            		github.com/NicholasDotSol/duality/x/dex/keeper/core.go:46 +0x1034
        	            	github.com/NicholasDotSol/duality/x/dex/keeper.msgServer.Deposit({{{0x244be50, 0x4000f5c8b0}, {0x242a618, 0x4000fde4d0}, {0x242a618, 0x0}, {{0x244be50, 0x4000f5c8b0}, 0x40001a02e0, {0x242a618, ...}, ...}, ...}}, ...)
        	            		github.com/NicholasDotSol/duality/x/dex/keeper/msg_server.go:40 +0x28c
        	            	github.com/NicholasDotSol/duality/x/dex/types._Msg_Deposit_Handler.func1({0x243ee40, 0x4004dc7170}, {0x1958960?, 0x4004d892c0})
        	            		github.com/NicholasDotSol/duality/x/dex/types/tx.pb.go:1038 +0x74
        	            	github.com/cosmos/cosmos-sdk/baseapp.(*MsgServiceRouter).RegisterService.func2.1({0x243ee40, 0x4004dc7140}, {0x4004e59ed8?, 0xd88dc0?}, 0x198b8c0?, 0x4004d8aac8)
        	            		github.com/cosmos/cosmos-sdk@v0.46.6/baseapp/msg_service_router.go:113 +0xc0
        	            	github.com/NicholasDotSol/duality/x/dex/types._Msg_Deposit_Handler({0x1992cc0?, 0x400011d220}, {0x243ee40, 0x4004dc7140}, 0x220f930, 0x4004dcaac0)
        	            		github.com/NicholasDotSol/duality/x/dex/types/tx.pb.go:1040 +0x134
        	            	github.com/cosmos/cosmos-sdk/baseapp.(*MsgServiceRouter).RegisterService.func2({{0x243edd0, 0x40001a6020}, {0x244da20, 0x4004dce380}, {{0xb, 0x0}, {0x4004c8e849, 0x7}, 0x25, {0xf30274, ...}, ...}, ...}, ...)
        	            		github.com/cosmos/cosmos-sdk@v0.46.6/baseapp/msg_service_router.go:126 +0x31c
        	            	github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).runMsgs(_, {{0x243edd0, 0x40001a6020}, {0x244da20, 0x4004dce380}, {{0xb, 0x0}, {0x4004c8e849, 0x7}, 0x25, ...}, ...}, ...)
        	            		github.com/cosmos/cosmos-sdk@v0.46.6/baseapp/baseapp.go:760 +0x464
        	            	github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).runTx(0x4000264000, 0x2, {0x4004b911e0, 0x146, 0x160})
        	            		github.com/cosmos/cosmos-sdk@v0.46.6/baseapp/baseapp.go:717 +0x8e0
        	            	github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).Simulate(0x10?, {0x4004b911e0?, 0x1?, 0x20?})
        	            		github.com/cosmos/cosmos-sdk@v0.46.6/baseapp/test_helpers.go:23 +0x30
        	            	github.com/cosmos/cosmos-sdk/x/auth/tx.txServer.Simulate({{{0x0, 0x0, 0x0}, {0x2459678, 0x40001bf4a0}, {0x0, 0x0}, {0x24442d8, 0x4000f5c8b0}, {0x2450c98, ...}, ...}, ...}, ...)
        	            		github.com/cosmos/cosmos-sdk@v0.46.6/x/auth/tx/service.go:120 +0x118
        	            	github.com/cosmos/cosmos-sdk/types/tx._Service_Simulate_Handler({0x18afd40?, 0x40011751e0}, {0x243ee40, 0x4004d944b0}, 0x4004d785f0, 0x0)
        	            		github.com/cosmos/cosmos-sdk@v0.46.6/types/tx/service.pb.go:910 +0x16c
        	            	github.com/cosmos/cosmos-sdk/baseapp.(*GRPCQueryRouter).RegisterService.func1({{0x243edd0, 0x40001a6020}, {0x244da20, 0x4004d8d140}, {{0xb, 0x0}, {0x4004c8e849, 0x7}, 0x25, {0xf30274, ...}, ...}, ...}, ...)
        	            		github.com/cosmos/cosmos-sdk@v0.46.6/baseapp/grpcrouter.go:85 +0x1a8
        	            	github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).handleQueryGRPC(0x4000264000, 0x4000cbb440, {{0x4002db3600, 0x149, 0x2c0}, {0x4004d68900, 0x23}, 0x25, 0x0})
        	            		github.com/cosmos/cosmos-sdk@v0.46.6/baseapp/abci.go:583 +0x124
        	            	github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).Query(0x4000264000, {{0x4002db3600, 0x149, 0x2c0}, {0x4004d68900, 0x23}, 0x25, 0x0})
        	            		github.com/cosmos/cosmos-sdk@v0.46.6/baseapp/abci.go:427 +0x78c
        	            	github.com/tendermint/tendermint/abci/client.(*localClient).QuerySync(0x4000d7b7a0, {{0x4002db3600, 0x149, 0x2c0}, {0x4004d68900, 0x23}, 0x0, 0x0})
        	            		github.com/tendermint/tendermint@v0.34.24/abci/client/local_client.go:256 +0x144
        	            	github.com/tendermint/tendermint/proxy.(*appConnQuery).QuerySync(0x40033cfa90?, {{0x4002db3600, 0x149, 0x2c0}, {0x4004d68900, 0x23}, 0x0, 0x0})
        	            		github.com/tendermint/tendermint@v0.34.24/proxy/app_conn.go:159 +0x5c
        	            	github.com/tendermint/tendermint/rpc/core.ABCIQuery(0x8?, {0x4004d68900?, 0x40001fe9c0?}, {0x4002db3600?, 0x4004d60b08?, 0x48e528?}, 0x2?, 0x2?)
        	            		github.com/tendermint/tendermint@v0.34.24/rpc/core/abci.go:20 +0x84
        	            	reflect.Value.call({0x178acc0?, 0x2211490?, 0x4004d61228?}, {0x199bdca, 0x4}, {0x4004d1f280, 0x5, 0x10ea370?})
        	            		reflect/value.go:584 +0x688
        	            	reflect.Value.Call({0x178acc0?, 0x2211490?, 0x2e5?}, {0x4004d1f280?, 0x4004d784b0?, 0x4004d65940?})
        	            		reflect/value.go:368 +0x90
        	            	github.com/tendermint/tendermint/rpc/jsonrpc/server.makeJSONRPCHandler.func1({0x243d590, 0x4004d23dd0}, 0x4004d74f00)
        	            		github.com/tendermint/tendermint@v0.34.24/rpc/jsonrpc/server/http_json_handler.go:108 +0xdbc
        	            	github.com/tendermint/tendermint/rpc/jsonrpc/server.handleInvalidJSONRPCPaths.func1({0x243d590?, 0x4004d23dd0?}, 0xffff8dfc7108?)
        	            		github.com/tendermint/tendermint@v0.34.24/rpc/jsonrpc/server/http_json_handler.go:140 +0x70
        	            	net/http.HandlerFunc.ServeHTTP(0x0?, {0x243d590?, 0x4004d23dd0?}, 0x10ed3f8?)
        	            		net/http/server.go:2109 +0x38
        	            	net/http.(*ServeMux).ServeHTTP(0x4000070800?, {0x243d590, 0x4004d23dd0}, 0x4004d74f00)
        	            		net/http/server.go:2487 +0x140
        	            	github.com/tendermint/tendermint/rpc/jsonrpc/server.maxBytesHandler.ServeHTTP({{0x2427020?, 0x400057fac0?}, 0x10?}, {0x243d590?, 0x4004d23dd0}, 0x4004d74f00)
        	            		github.com/tendermint/tendermint@v0.34.24/rpc/jsonrpc/server/http_server.go:256 +0x138
        	            	github.com/tendermint/tendermint/rpc/jsonrpc/server.RecoverAndLogHandler.func1({0x243dbf0?, 0x4004d521c0}, 0x4004d74f00)
        	            		github.com/tendermint/tendermint@v0.34.24/rpc/jsonrpc/server/http_server.go:229 +0x2b4
        	            	net/http.HandlerFunc.ServeHTTP(0x0?, {0x243dbf0?, 0x4004d521c0?}, 0x8587c4?)
        	            		net/http/server.go:2109 +0x38
        	            	net/http.serverHandler.ServeHTTP({0x242b950?}, {0x243dbf0, 0x4004d521c0}, 0x4004d74f00)
        	            		net/http/server.go:2947 +0x2cc
        	            	net/http.(*conn).serve(0x40048c99a0, {0x243ee40, 0x400051e300})
        	            		net/http/server.go:1991 +0x544
        	            	created by net/http.(*Server).Serve
        	            		net/http/server.go:3102 +0x43c
        	            	: panic [cosmos/cosmos-sdk@v0.46.6/baseapp/recovery.go:69] With gas wanted: '0' and gas used: '50665' : unknown request
        	            	Usage:
        	            	  dualityd tx dex deposit [receiver] [token-a] [token-b] [list of amount-0] [list of amount-1] [list of tick-index] [list of fee]  [flags]
        	            	
        	            	Flags:
        	            	  -a, --account-number uint      The account number of the signing account (offline mode only)
        	            	  -b, --broadcast-mode string    Transaction broadcasting mode (sync|async|block) (default "sync")
        	            	      --dry-run                  ignore the --gas flag and perform a simulation of a transaction, but don't broadcast it (when enabled, the local Keybase is not accessible)
        	            	      --fee-account string       Fee account pays fees for the transaction instead of deducting from the signer
        	            	      --fees string              Fees to pay along with transaction; eg: 10uatom
        	            	      --from string              Name or address of private key with which to sign
        	            	      --gas string               gas limit to set per-transaction; set to "auto" to calculate sufficient gas automatically (default 200000)
        	            	      --gas-adjustment float     adjustment factor to be multiplied against the estimate returned by the tx simulation; if the gas limit is set manually this flag is ignored  (default 1)
        	            	      --gas-prices string        Gas prices in decimal format to determine the transaction fee (e.g. 0.1uatom)
        	            	      --generate-only            Build an unsigned transaction and write it to STDOUT (when enabled, the local Keybase is not accessible)
        	            	  -h, --help                     help for deposit
        	            	      --keyring-backend string   Select keyring's backend (os|file|kwallet|pass|test|memory) (default "test")
        	            	      --keyring-dir string       The client Keyring directory; if omitted, the default 'home' directory will be used
        	            	      --ledger                   Use a connected Ledger device
        	            	      --node string              <host>:<port> to tendermint rpc interface for this chain (default "tcp://localhost:26657")
        	            	      --note string              Note to add a description to the transaction (previously --memo)
        	            	      --offline                  Offline mode (does not allow any online functionality
        	            	  -o, --output string            Output format (text|json) (default "json")
        	            	  -s, --sequence uint            The sequence number of the signing account (offline mode only)
        	            	      --sign-mode string         Choose sign mode (direct|amino-json), this is an advanced feature
        	            	      --timeout-height uint      Set a block timeout height to prevent the tx from being committed past a certain height
        	            	  -y, --yes                      Skip tx broadcasting prompt confirmation
        	            	
        	            	Global Flags:
        	            	      --chain-id string     The network chain ID (default "duality")
        	            	      --home string         directory for config and data (default "/home/heighliner/.duality")
        	            	      --log_format string   The logging format (json|plain) (default "plain")
        	            	      --log_level string    The logging level (trace|debug|info|warn|error|fatal|panic) (default "info")
        	            	      --trace               print out full stack trace on errors
```

## To Replicate the Issue

You will need to build the Duality Docker image via [heighliner](https://github.com/strangelove-ventures/heighliner)

After cloning and compiling heighliner, checkout this branch https://github.com/duality-labs/duality/tree/justin/ibc-swap

From the root directory of the duality repo you can run the following command to build the Docker image
`heighliner build -c duality --local`

With the Docker image built you can now run the test case with
`cd ibctest && go test -race -v -run TestIBCSwapMiddleware_Success .`

**Comment by jtieri** [comment]
A patch for this was merged as I was typing out the issue. Closed by #290

**Comment by NicholasDotSol** [comment]
Thanks @jtieri!
fixed by #290, closing this
